client/context/query.go

```diff
@@ -72,22 +74,64 @@ func (ctx CLIContext) GetAccount(address []byte) (auth.Account, error) {
 }
 
 // GetFromAddress returns the from address from the context's name.
-func (ctx CLIContext) GetFromAddress() (from sdk.AccAddress, err error) {
-	if ctx.FromAddressName == "" {
-		return nil, errors.Errorf("must provide a from address name")
+func (ctx CLIContext) GetFromAddress() (sdk.AccAddress, error) {
+	if ctx.fromAddress != nil {
+		return ctx.fromAddress, nil
 	}
 
-	keybase, err := keys.GetKeyBase()
+	err := ctx.populateFromFields()
 	if err != nil {
 		return nil, err
 	}
 
-	info, err := keybase.Get(ctx.FromAddressName)
+	return ctx.fromAddress, nil
+}
+
+func (ctx CLIContext) GetFromName() (string, error) {
+	if ctx.fromName != "" {
+		return ctx.fromName, nil
+	}
+
+	err := ctx.populateFromFields()
 	if err != nil {
-		return nil, errors.Errorf("no key for: %s", ctx.FromAddressName)
+		return "", err
+	}
+
+	return ctx.fromName, nil
+}
+
+func (ctx CLIContext) populateFromFields() error {
```

Comment:
Are you sure this logic is correct? I have not been able to get this to work (`gaiacli send --from <addr1> --to <addr2> ...`)

Reply:
Yep - note that you'll have to create a new key for the correct data to appear in the database.

Reply:
To be clear: the way I've written it, an additional key (as in key-value store) is written to LevelDB to look up keys (as in private keys) by address. If that KV-store key doesn't exist, you won't be able to pass a public key in `--from`.

Reply:
@mslipper I created a fresh local testnet, added all of my node's keys to a single node's key DB and still could not get this to work. I've tried with both recovered keys and with newly created keys.

Reply:
Were you able to get it working @alexanderbez ? If you could post the command you're using and the error you get I can debug.